Add gradle plugin for downloading jdk

rjernst opened the pull request:
We currently download 3 variants of the same version of the jdk for
bundling into the distributions. Additionally, the vagrant images do
their own downloading. This commit moves the jdk downloading into a
utility gradle plugin. This will be used in a future PR by the packaging
tests.

The new plugin exposes a "usesJdk" method on tasks. Once the jdk version
and platform are set for a task, a "jdkHome" property exists which may
be used to get the extracted jdk directory at runtime.

rjernst committed:
Add gradle plugin for downloading jdk

We currently download 3 variants of the same version of the jdk for
bundling into the distributions. Additionally, the vagrant images do
their own downloading. This commit moves the jdk downloading into a
utility gradle plugin. This will be used in a future PR by the packaging
tests.

The new plugin exposes a "usesJdk" method on tasks. Once the jdk version
and platform are set for a task, a "jdkHome" property exists which may
be used to get the extracted jdk directory at runtime.

elasticmachine commented:
Pinging @elastic/es-core-infra

rjernst committed:
simplify

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/JdkDownloadPlugin.java:
```diff
@@ -0,0 +1,187 @@
+                    if (ALLOWED_PLATFORMS.contains(platform) == false) {
+                        throw new IllegalArgumentException("platform must be one of " + ALLOWED_PLATFORMS);
+                    }
+                    ExtraPropertiesExtension extraProperties = task.getExtensions().findByType(ExtraPropertiesExtension.class);
```
nit: could be moved up so it's repeated on line 60

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/JdkDownloadPlugin.java:
```diff
@@ -0,0 +1,187 @@
+                    }
+                    ExtraPropertiesExtension extraProperties = task.getExtensions().findByType(ExtraPropertiesExtension.class);
+                    if (extraProperties.has("jdkHome")) {
+                        throw new IllegalArgumentException("jdk version already set for task");
```
Maybe include in the exception what version was set previously

rjernst replied:
This is actually difficult to figure out, since the version is not directly stored anywhere, it is only part of various configuration names etc.

atorok replied:
maybe `extraProperties.get("jdkHome")` would help ?

rjernst replied:
It is just an arbitrary object, and can't be resolved to toString until runtime, while this is at configuration time.

atorok replied:
ok, let's leave it for now, we can revisit if it becomes too painful to debug

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/JdkDownloadPlugin.java:
```diff
@@ -0,0 +1,187 @@
+                    }
+
+                    // ensure a root level jdk download task exists
+                    setupRootJdkDownload(project.getRootProject(), platform, version);
```
Could we set up a single task to download all as we do for test clusters rather than one for  each version ? Having fewer tasks will make configuration faster and I find it easier to follow what's going on. 
It could also be a lazily created task that's always set up and  then have the tasks that require it depend on it

atorok replied:
I might be missing it, but is `task` made dependent on this ?

rjernst replied:
I do not think we should have a single configuration for all jdks we download. That means resolving the configuration triggers download of all of them. Additionally, the way this works is to be able to resolve a configuration to a single file.

> is task made dependent on this ?

The root project does the extraction, and makes the extracted dir available as an artifact of the root project. The tasks in subprojects add a normal dependency on that "artifact", through a separate configuration in the subproject, and the task then depends on that local subproject configuration.

atorok replied:
I wasn't referring to having a single configuration, we actually need those to be separate to be able to have different versions, but we can extract all in the same task, having multiple tasks for it doesn't buy us anything.

rjernst replied:
Having a single task would mean the task depends on all those configurations, thus still resolving all the artifacts at the same time. We need a 1-1 mapping to ensure one extraction is done.

atorok replied:
Is that all that bad?   The configurations are cached, and the task has up to date checks .  For most things this all the JDKs will be needed anyway due to the number of tests that depend on the JDK.

rjernst replied:
> For most things this all the JDKs will be needed anyway

I don't think that is true. When I run a packaging test for a single distro, I need one jdk, not many, and certainly not the ones from non linux.

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/JdkDownloadPlugin.java:
```diff
@@ -0,0 +1,187 @@
+                ivyRepo.setName(repoName);
+                ivyRepo.setUrl("https://download.java.net");
+                ivyRepo.patternLayout(layout -> {
+                    layout.artifact("java/GA/jdk" + jdkMajor + "/" + jdkBuild + "/GPL/openjdk-[revision]_[module]-x64_bin.[ext]");
```
we could add something like: 
```
spec.content(c-> c.includeGroup("org.elasticsearch.jdk-downloader."));
```
And use "org.elasticsearch.jdk-downloader" as a group when downloading the dependency. 
It's a good practice for repositories confirmed by plugins.

rjernst replied:
Sorry I don't follow. The group is completely arbitrary and internal since this is a fake ivy repo.

atorok replied:
But the repository is not, so all dependencies could potentially be checked against it. 
I just suggested `org.elasticsearch.jdk-downloader` to make it more specific, or make it "ours" since it's not really used in the pattern layout. It would work the same way with "jdk".

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/JdkDownloadPlugin.java:
```diff
@@ -0,0 +1,187 @@
+        final ConfigurationContainer configurations = rootProject.getConfigurations();
+        String remoteConfigName = configName("openjdk", version, platform);
+        String localConfigName = configName("localjdk", version, platform);
+        Configuration jdkConfig = configurations.findByName(remoteConfigName);
```
I think we repeat this pattern more and more. I think all of these collections are `NamedDomainObjectCollection`s so I think we could add a `Boilerplate.fondOrCreate(NamedDomainObjectCollection collection, String name)` method to make it easier on the eyes.

rjernst replied:
That may be useful, but I think this PR is big enough to warrant that as a separate discussion.

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/JdkDownloadPlugin.java:
```diff
@@ -0,0 +1,187 @@
+        if (extension.equals("zip")) {
+            fileGetter = () -> rootProject.zipTree(jdkArchiveGetter.get());
+        } else {
+            fileGetter = () -> rootProject.tarTree(rootProject.getResources().gzip(jdkArchiveGetter.get()));
```
nit: no need to be explicit about gzip `tarTree` should figure that out.

rjernst replied:
I prefer explicit. It is easier to reason about, and doesn't leave the possibility for magic to change and cause problems.

atorok commented on buildSrc/src/test/java/org/elasticsearch/gradle/JdkDownloadPluginTests.java:
```diff
@@ -0,0 +1,104 @@
+        project.getTasks().create("example1", task -> usesJdk(task, "11.0.2+33", "linux"));
+        int numTasks = rootProject.getTasks().size();
+        project.getTasks().create("example2", task -> usesJdk(task, "11.0.2+33", "linux"));
+        assertThat(rootProject.getTasks().size(), equalTo(numTasks));
```
how come this isn't off by one  due to the one additional task created on the line above ?

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/JdkDownloadPlugin.java:
```diff
@@ -0,0 +1,187 @@
+                            return configurations.getByName(configurationName).getSingleFile().toString();
+                        }
+                    };
+                    extraProperties.set("jdkHome", jdkHomeGetter);
```
Rather than registering this in the `ExtraProeprtiesExtension` I think we should define a custom one, has a reference to configurations and does what the above `jdkHomeGetter` would do for this property. 
The advantage is that we get our own namespace, so when build scripts refer to this they do it with `jdkDownloader.jdkHome` and `jdkDownlaoder.usesJDK` making it more obvious where these come from.

rjernst replied:
Sure, I can try to add an extension.

rjernst replied:
I've moved jdkHome to an extension, but I would like to keep usesJdk as it is, because I don't want the extra level of indirection (I think it makes it less like a dsl).

rjernst replied:
Actually I had to revert that even. Apparently an extension must be an abstract type/interface, which means for jdkHome we would need an extra level of indirection there too. So I've kept it as an ext property for now; I'm happy to discuss cleaning this up as a followup.

rjernst committed:
review feedback

rjernst committed:
back to ext props

rjernst committed:
use an "extension", but not a real extension

rjernst committed:
fix tests

rjernst committed:
alternative

rjernst commented:
@atorok I wasn't happy with usesJdk on every task, so I rethought this and have pushed an alternative approach. The new code has a `jdks` container extension on the project where each jdk used has a name. Then depending on the jdk in a task is as simple as `dependsOn jdks.myjdk`, and using it eg for cli args is `args jdks.myjdk.homeDir` and to copy the files `from jdks.myjdk.files`.

rjernst commented:
Note that I need to update the tests with the new approach, but wanted to get this up here for you to look at if you have time today.

rjernst committed:
tests

rjernst committed:
fix checkstyle

rjernst commented:
@elasticmachine run elasticsearch-ci/1

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/JdkDownloadPlugin.java:
```diff
@@ -0,0 +1,165 @@
+            configurations.create(localConfigName);
+        }
+        String extension = platform.equals("windows") ? "zip" : "tar.gz";
+        String jdkDep = "jdk:" + (platform.equals("darwin") ? "osx" : platform) + ":" + jdkVersion + "@" + extension;
```
I think I'd move the dependency notation logic into `Jdk` to simplify the plugin code a bit. I don't have strong feelings here, but sometimes I find moving some of these details out of the plugin makes the logic a bit easier to reason about.

rjernst replied:
My intent was to have the real "logic" in the plugin. `Jdk` is just a container for information.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/JdkDownloadPlugin.java:
```diff
@@ -0,0 +1,165 @@
+        String jdkDep = "jdk:" + (platform.equals("darwin") ? "osx" : platform) + ":" + jdkVersion + "@" + extension;
+        rootProject.getDependencies().add(configName("openjdk", version, platform), jdkDep);
+
+        // add task for extraction
```
I _think_ it could be possible to do this as an artifact transform at some point. Starting in Gradle 5.3 these are also cacheable. Perhaps at a `TODO` for us to revisit this at some point.

rjernst replied:
Added a todo.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/JdkDownloadPlugin.java:
```diff
@@ -0,0 +1,165 @@
+
+    private static void setupRootJdkDownload(Project rootProject, String platform, String version) {
+        String extractTaskName = "extract" + capitalize(platform) + "Jdk" + version;
+        // NOTE: this is *horrendous*, but seems to be the only way to check for the existence of a registered task
```
Alternatively, could we look for the associated configuration instead of the extract task? We are just trying to determine if we've seen this version before so either method would work just as well.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/JdkDownloadPlugin.java:
```diff
@@ -0,0 +1,165 @@
+        rootProject.getDependencies().add(configName("openjdk", version, platform), jdkDep);
+
+        // add task for extraction
+        int rootNdx = platform.equals("darwin") ? 2 : 1;
```
Is this the case for all osx jdk distributions?

rjernst replied:
At least for the openjdk ones we download, yes it is.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/JdkDownloadPlugin.java:
```diff
@@ -0,0 +1,165 @@
+        int rootNdx = platform.equals("darwin") ? 2 : 1;
+        Action<CopySpec> removeRootDir = copy -> {
+            // remove extra unnecessary directory levels
+            copy.eachFile((FileCopyDetails details) -> {
```
Is this cast to `FileCopyDetails` necessary?

mark-vieira reviewed (changes requested):
Some misc comments, otherwise this looks great.

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/Jdk.java:
```diff
@@ -0,0 +1,114 @@
+public class Jdk implements Buildable {
+
+    static final Pattern VERSION_PATTERN = Pattern.compile("(\\d+)(\\.\\d+\\.\\d+)?\\+(\\d+)");
+    private static final List<String> ALLOWED_PLATFORMS = Collections.unmodifiableList(Arrays.asList("linux", "windows", "darwin"));
```
We have something similar: `org.elasticsearch.gradle.OS`

rjernst replied:
I don't think an enum has any advantages here. In fact, it would require more unnecessary code to convert between the capitalized enum entities and the lowercase platform names we need here.

atorok commented on buildSrc/src/testKit/jdk-download/subproj/build.gradle:
```diff
@@ -0,0 +1,41 @@
+
+String fakeJdkVersion = Objects.requireNonNull(System.getProperty('tests.jdk_version'))
+jdks {
+    linux_jdk {
```
nit: could we remove `_jdk` from the name to avoid redundancy ? And keep that as the naming format in general.

rjernst replied:
This is just an example in a test, but I removed it.

atorok commented:
> @atorok I wasn't happy with usesJdk on every task, so I rethought this and have pushed an alternative approach. The new code has a `jdks` container extension on the project where each jdk used has a name. Then depending on the jdk in a task is as simple as `dependsOn jdks.myjdk`, and using it eg for cli args is `args jdks.myjdk.homeDir` and to copy the files `from jdks.myjdk.files`.

@rjernst  ++ I like the approach very much !

You might be able to add some additional magic with `Object#asType(Class)`. 
I'm not 100% sure if that helps or just creates more confusion, but you would be able to always pass just `jdks.myjdk`.

rjernst commented on buildSrc/src/main/java/org/elasticsearch/gradle/Jdk.java:
```diff
@@ -0,0 +1,114 @@
+import java.util.List;
+import java.util.regex.Pattern;
+
+public class Jdk implements Buildable {
```
I'm happy to add this. Can you explain what benefits it would give us here?

mark-vieira replied:
It's just a bit more explicit and I believe I recall some Gradle APIs that handle a `Named` explicitly.

rjernst commented on buildSrc/src/main/java/org/elasticsearch/gradle/Jdk.java:
```diff
@@ -0,0 +1,114 @@
+        return configuration;
+    }
+
+    public Object getHomeDir() {
```
It is used. See the example "subproj" used in the integ test.

mark-vieira replied:
Gotcha, IntelliJ didn't catch that.

rjernst committed:
Merge branch 'master' into jdk_download

rjernst commented on buildSrc/src/main/java/org/elasticsearch/gradle/JdkDownloadPlugin.java:
```diff
@@ -0,0 +1,165 @@
+        if (rootProject.getRepositories().findByName(repoName) == null) {
+            rootProject.getRepositories().ivy(ivyRepo -> {
+                ivyRepo.setName(repoName);
+                ivyRepo.setUrl("https://download.java.net");
```
This would add a considerable load to our infrastructure. We may want to still consider it, but I think it should be separate from this PR, which is just to consolidate existing code, which already uses this url.

mark-vieira replied:
I still have a bit of a concern with this approach of registering so many repositories. Is there anyway perhaps that we could construct the dependency notation so that we can avoid creating a repo per version? I'm not sure this is possible, or if what I see here even works. For example, the download link I see [here](https://jdk.java.net/12/) includes the artifact hash in the download URL.

rjernst replied:
I don't think it is possible here. The reason we embed this information in the pattern is we only have so many elements we can substitute via the maven/ivy coordinates. 

The hash included in the download link is something that changed very recently. There are 2 versions registered here because versions before 12.0.1 still use the old pattern. I could definitely decide which one to register based on the version, but in this PR I'm only trying to move the logic we already have in the build for jdk downloading. We can certainly look at improving this either through reducing repos or moving to using the jvm catalog, but I think these should be discussed separately as followups to this PR.

mark-vieira replied:
Since we're only registering all these repos to the root project it's probably ok, as it won't affect dependency solution of subprojects. Let's make sure we create an issue to follow up here. I assume if the recent version download links have changed this is going to be a problem for fetching newer JDK versions, no?

rjernst replied:
The older links didn't change. Oracle only updated the url pattern for newer jdk releases. This is why we have both patterns here.

mark-vieira replied:
I think I'm missing something. I only see a single artifact pattern definition. Is this going to work with newer JDKs that require the artifact hash to be part of the URL?

rjernst replied:
We register 2 ivy repositories, one with the older pattern, and one with the hash. Are you sure you are looking at the latest commit?

mark-vieira replied:
I definitely was not. It seems GitHub was filtering out the latest commits for me. Thanks GH!

mark-vieira replied:
Given that we know the version when registering the repo can we be smart about it and avoid registering both for all versions?

mark-vieira replied:
Or simpler yet, if the artifact version notation includes a hash, use the hash repo, otherwise use the legacy one.

rjernst replied:
Sure, I will do that in a followup.

rjernst committed:
address feedback

rjernst committed:
fix oops

rjernst committed:
make jdk useable as project.files and lazy string

rjernst commented:
> You might be able to add some additional magic with Object#asType(Class).

Thanks for the idea! While I didn't use `asType`, I was able to eliminate `homeDir` and `fileCollection` by making `Jdk` implement toString and Iterable<File>.

rjernst commented:
@atorok @mark-vieira I believe I have addressed all your feedback.

alpar-t reviewed (approved):
LGTM

We might still need an `asType(File)` to be able to pass a JDK to an API that expects a File, but we can add that whenever we need it.

mark-vieira commented:
Yeah, I'm on the fence on using `asType()`. It still would require you to do something like `myJdk as File` which I'm not certain is significantly better than `myJdk.file`. I'm generally in favor of using the least amount of Groovy magic as possible unless it makes the DSL significantly cleaner.

jaymode committed:
Update TLS ciphers and protocols for JDK 11 (#41808)

* Update TLS ciphers and protocols for JDK 11 (#41385)

This commit updates the default ciphers and TLS protocols that are used
after the minimum supported JDK is JDK 11. The conditionals around
TLSv1.3 and 256-bit cipher support have been removed. JDK 11 no longer
requires an unlimited JCE policy file for 256 bit cipher support and
TLSv1.3 is supported in JDK 11+. New cipher support has been introduced
in the newer JDK versions as well. The ciphers are ordered with PFS
ciphers being most preferred, then AEAD ciphers, and finally those with
mainstream hardware support.

* Fixes for TLSv1.3 on JDK11

* fix for JDK-8212885

romseygeek committed:
Correct spelling of MockLogAppender.PatternSeenEventExpectation (#41893)

The class was called PatternSeenEventExcpectation. This commit
is a straight class rename to correct the spelling.

rjernst committed:
Merge branch 'master' into jdk_download

rjernst commented:
@elasticmachine run elasticsearch-ci/1

rjernst commented:
@elasticmachine run elasticsearch-ci/packaging

rjernst committed:
Merge branch 'master' into jdk_download

rjernst committed:
remove bad files

rjernst commented:
@elasticmachine run elasticsearch-ci/1

rjernst committed:
Merge branch 'master' into jdk_download
